Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm

```diff
@@ -51,15 +42,34 @@ void DispatchAsync(const dispatch_queue_t queue, std::function<void()>&& work) {
   });
 }
 
-void DispatchSync(const dispatch_queue_t queue, Executor::Operation work) {
+void DispatchSync(const dispatch_queue_t queue, std::function<void()> work) {
+  FIREBASE_ASSERT_MESSAGE(
+      StringViewFromDispatchLabel(dispatch_queue_get_label(queue)) !=
+          StringViewFromDispatchLabel(
+              dispatch_queue_get_label(dispatch_get_main_queue())),
+      "Calling dispatch_sync on the main queue will lead to a deadlock.");
```

Comment:
`dispatch_sync(dispatch_get_main_queue(), ^{})` works just fine:

```c++
#import <Foundation/Foundation.h>

#include <cstdlib>
#include <iostream>

int main() {
  dispatch_queue_t queue = dispatch_get_main_queue();
  dispatch_queue_t secondary = dispatch_queue_create("secondary", nullptr);
  dispatch_async(secondary, ^{
    dispatch_sync(queue, ^{
      std::cout << "Hello world!\n";
    });
    exit(0);
  });

  dispatch_main();
}
```
What doesn't work is `dispatch_sync` or `dispatch_async` on the main queue if `dispatch_main` hasn't been called yet. That's called normally as a part of the run loop of iOS/macOS applications. Perhaps the problem is that the GoogleTest-based tests don't do that?

Reply:
You're right; I misunderstood the issue I was experiencing. The real problem was calling `dispatch_sync` on the current queue, which happened to be the main queue (from the docs, _"Calling this function and targeting the current queue results in deadlock"_). Reworked the assertion, PTAL.